File: apps/ios/ScribbleFit/LocalPackages/Features/FeatureExercises/Sources/FeatureExercises/Domain/UseCases/GetExerciseHistoryUseCase.swift
import Foundation
import CoreModel
import CoreCommon

/**
 * Use case to fetch the complete history of a specific exercise.
 */
@MainActor
public final class GetExerciseHistoryUseCase {
    private let exerciseRepository: ExerciseRepository
    private let formatExerciseSummaryUseCase: FormatExerciseSummaryUseCase
    
    public init(
        exerciseRepository: ExerciseRepository,
        formatExerciseSummaryUseCase: FormatExerciseSummaryUseCase
    ) {
        self.exerciseRepository = exerciseRepository
        self.formatExerciseSummaryUseCase = formatExerciseSummaryUseCase
    }
    
    public func execute(
        exerciseName: String,
        weightUnit: WeightUnit
    ) async throws -> [ExerciseHistorySession] {
        let history = try await exerciseRepository.getExercises(query: exerciseName)
            .filter { $0.canonicalName.lowercased() == exerciseName.lowercased() }
            .sorted(by: { $0.createdAt > $1.createdAt })
        
        if history.isEmpty { return [] }
        
        // Calculate max weight and volume across all time to identify PBs
        let allTimeMaxWeight = history.flatMap { $0.sets }.compactMap { $0.weight }.max() ?? 0.0
        let allTimeMaxVolume = history.map { exercise in
            exercise.sets.reduce(Float(0.0)) { $0 + Calculations.calculateVolume(weight: $1.weight, reps: $1.reps) }
        }.max() ?? 0.0
        
        return history.map { exercise in
            let sessionVolume = exercise.sets.reduce(Float(0.0)) { $0 + Calculations.calculateVolume(weight: $1.weight, reps: $1.reps) }
            let sessionMaxWeight = exercise.sets.compactMap { $0.weight }.max() ?? 0.0
            
            let isPB = (sessionMaxWeight >= allTimeMaxWeight && allTimeMaxWeight > 0.0) ||
                       (sessionVolume >= allTimeMaxVolume && allTimeMaxVolume > 0.0)
            
            return ExerciseHistorySession(
                exercise: exercise,
                totalVolume: sessionVolume,
                maxWeight: sessionMaxWeight,
                summary: formatExerciseSummaryUseCase.execute(exercise: exercise, weightUnit: weightUnit),
                isPersonalBest: isPB,
                scribbleId: exercise.scribbleId ?? UUID()
            )
        }
    }
}

File: apps/android/feature/exercises/domain/src/main/java/com/scribblefit/feature/exercises/domain/usecase/GetExerciseHistoryUseCase.kt
package com.scribblefit.feature.exercises.domain.usecase

import com.scribblefit.core.common.Calculations
import com.scribblefit.core.config.domain.Weight
import com.scribblefit.feature.exercises.domain.ExerciseRepository
import com.scribblefit.core.model.ExerciseHistorySession
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.withContext

/**
 * Use case to fetch the complete history of a specific exercise.
 */
class GetExerciseHistoryUseCase(
    private val exerciseRepository: ExerciseRepository,
    private val formatExerciseSummaryUseCase: FormatExerciseSummaryUseCase,
    private val coroutineDispatcher: CoroutineDispatcher
) {
    suspend operator fun invoke(
        exerciseName: String,
        weightUnit: Weight
    ): Result<List<ExerciseHistorySession>> = withContext(coroutineDispatcher) {
        runCatching {
            val history = exerciseRepository.getExercisesByName(exerciseName)
                .sortedByDescending { it.createdAt }

            if (history.isEmpty()) return@runCatching emptyList()

            // Calculate max weight and volume across all time to identify PBs
            val allTimeMaxWeight = history.flatMap { it.sets }.maxOfOrNull { it.weight ?: 0f } ?: 0f
            val allTimeMaxVolume = history.maxOfOrNull { ex ->
                ex.sets.sumOf { set -> Calculations.calculateVolume(set.weight, set.reps).toDouble() }.toFloat()
            } ?: 0f

            history.map { exercise ->
                val sessionVolume = exercise.sets.sumOf { set ->
                    Calculations.calculateVolume(set.weight, set.reps).toDouble()
                }.toFloat()
                
                val sessionMaxWeight = exercise.sets.maxOfOrNull { it.weight ?: 0f } ?: 0f

                // A session is a PB if it contains the all-time max weight or achieves all-time max volume
                // Note: In a real app, you might want more nuanced PB logic (e.g. per rep range)
                val isPB = (sessionMaxWeight >= allTimeMaxWeight && allTimeMaxWeight > 0f) || 
                           (sessionVolume >= allTimeMaxVolume && allTimeMaxVolume > 0f)

                // We need the scribbleId to navigate back to the original workout.
                // The repository should provide this. 
                // Currently Exercise model doesn't have scribbleId, but ExerciseEntity does.
                // I'll check if I need to update the Exercise domain model or if I can get it another way.
                
                ExerciseHistorySession(
                    exercise = exercise,
                    totalVolume = sessionVolume,
                    maxWeight = sessionMaxWeight,
                    summary = formatExerciseSummaryUseCase(exercise, weightUnit),
                    isPersonalBest = isPB,
                    scribbleId = exercise.scribbleId
                )
            }
        }
    }
}
